### PR TITLE
GLM user example with custom coefficient optimization method

### PR DESCRIPTION
**Pull request by Noy-Maimon**
This example is a result of the KSM workshop, demonstrating how to configure a GLM model using NVFLARE and rhino's FCP.

### CODE REVIEW

**Review by yuvalrhino — CHANGES_REQUESTED** [review]
Some comments/questions, but overall this looks awesome!

File: examples/nvflare/regression-glm-coeff/custom/epi_aggregation_helper.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (Noy-Maimon) [reply examples/nvflare/regression-glm-coeff/custom/epi_aggregation_helper.py:1]: Yeah it is based on an [nvflare file ](https://github.com/NVIDIA/NVFlare/blob/2.3/nvflare/app_common/aggregators/weighted_aggregation_helper.py)
So leaving as is

File: examples/nvflare/regression-glm-coeff/custom/coeff_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+            raise ValueError("The first guess on the deviance function "
+                             "returned a nan.  This could be a boundary "
+                             " problem and should be reported.")
+        return glm, lin_pred, mu
+
+    @staticmethod
+    def _site_irls_iteration(site_info, logger_warnings):
+        """
+        Perform one iteration of the IRLS algorithm.
+        This call is done once per site and runs locally at each site.
```
Comment (yuvalrhino) [inline examples/nvflare/regression-glm-coeff/custom/coeff_optimizer.py:144]:
```suggestion
        This method is called during each iteration of IRLS and runs locally at each site.
```

File: examples/nvflare/regression-glm-coeff/custom/coeff_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+class IRLS(CoeffOptimizer):
+
+    def name(self):
+        return "IRLS"
+
+    @staticmethod
+    def _site_irls_initialization(glm, start_params=None):
+        """
+        Initialize the model for the IRLS algorithm.
+        This call is done once per site and runs locally at each site.
```
Comment (yuvalrhino) [inline examples/nvflare/regression-glm-coeff/custom/coeff_optimizer.py:120]:
```suggestion
        This method is called once per site and runs locally at each site.
```

File: examples/nvflare/regression-glm-coeff/custom/coeff_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+            'A': A,
+            'B': B,
+        }
+
+        return ols_params
+
+    def get_local_coeffs(self, current_round, np_data, formula, offset, family_class, logger_warnings, data=None, data_y=None, data_x=None, site_info=None):
+        """
+        Calculate the local coefficients, and modifies the np_data object with the results.
+        Runs locally at each site.
```
Comment (yuvalrhino) [inline examples/nvflare/regression-glm-coeff/custom/coeff_optimizer.py:206]:
```suggestion
        This method is called during each iteration of IRLS and runs locally at each site.
```

File: examples/nvflare/regression-glm-coeff/custom/epi_aggregation_helper.py (hunk 1 above)
Comment (yuvalrhino) [inline examples/nvflare/regression-glm-coeff/custom/epi_aggregation_helper.py:1]:
```suggestion
# Copyright (c) 2023, Rhino HealthTech, Inc.
# Original file modified by Rhino Health to adapt it to the Rhino Health Federated Computing Platform.

# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```

File: examples/nvflare/regression-glm-coeff/custom/epi_aggregator.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yuvalrhino) [inline examples/nvflare/regression-glm-coeff/custom/epi_aggregator.py:1]:
```suggestion
# Copyright (c) 2023, Rhino HealthTech, Inc.
# Original file modified by Rhino Health to adapt it to the Rhino Health Federated Computing Platform.

# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```

File: examples/nvflare/regression-glm-coeff/custom/glm_model_persistor.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yuvalrhino) [inline examples/nvflare/regression-glm-coeff/custom/glm_model_persistor.py:1]:
```suggestion
# Copyright (c) 2023, Rhino HealthTech, Inc.
# Original file modified by Rhino Health to adapt it to the Rhino Health Federated Computing Platform.

# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```

File: examples/nvflare/regression-glm-coeff/custom/regression_trainer.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yuvalrhino) [inline examples/nvflare/regression-glm-coeff/custom/regression_trainer.py:1]:
```suggestion
# Copyright (c) 2023, Rhino HealthTech, Inc.
# Original file modified by Rhino Health to adapt it to the Rhino Health Federated Computing Platform.

# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```

File: examples/nvflare/regression-glm-coeff/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+USER localuser
+
+# !! EDIT THIS: Copy the needed local files (code and otherwise) into the container work directory.
+# Note: For directories, have a separate COPY command for each top-level directory.
+COPY --chown=$UID:$GID config ./config
+COPY --chown=$UID:$GID custom ./custom
+COPY --chown=$UID:$GID infer.py ./infer.py
+
+# !! EDIT THIS: Ensure files are executable.
+RUN chmod +x ./infer.py
```
Comment (yuvalrhino) [inline examples/nvflare/regression-glm-coeff/Dockerfile:52]: Do we need an infer.py in this example? Or can we remove this line and the (empty) infer.py file?

**Review by yuvalrhino — CHANGES_REQUESTED** [review]
Thanks for making the changes! 
I've updated the notebook a bit to adapt it to the format we use in other example notebooks including markdown, etc.
In any files that contain the NVIDIA copyright that we have edited in any way, we should add a Rhino Health copyright at the top as well (any files that we took as is from NVIDIA can be left with just the NVIDIA copyright, and any files that aren't from NVIDIA at all don't need our copyright since we have a default copyright in the license file in the repo). Could you please review and make sure this is indeed the case?

File: examples/nvflare/regression-glm-coeff/Dockerfile (hunk 8 above)
Comment (Noy-Maimon) [reply examples/nvflare/regression-glm-coeff/Dockerfile]: Removed (+verified its not breaking anything)

**Review by yuvalrhino — APPROVED** [review]

**Pull request merged**